# Capstone Demo Notebook — under the hood of the referral platform

This notebook exercises the parts of `healthcare_capstone` that a click-through
UI demo cannot show: raw JWT tokens and their failure modes, RBAC decisions at
the permission-table level, the LangGraph referral workflow's actual state and
human-in-the-loop pauses, live MCP protocol calls (tools **and** resources
**and** prompts) against real running MCP servers, structured JSON logs tied
together by a correlation id, rate limiting, refresh-token reuse detection,
outbox events fanning out to a live SSE stream, and the conversational
assistant's role-scoped tool access.

Every section runs real project code — the same node functions, the same
route handlers, the same `ROLE_PERMISSIONS`/`ROLE_TOOL_ALLOWLIST` tables the
running app uses — not a re-implementation. The only substitutions are
infrastructure, not logic:

- **Postgres → in-memory SQLite**, for both the app's own tables and the
  LangGraph checkpointer (`AsyncSqliteSaver`, the same one `tests/conftest.py`
  uses for the exact same reason).
- **Real MCP sockets for the mocked external systems → the same
  `tests/agent_fakes.fake_get_tools` double the test suite uses**, which
  still round-trips through the real mock FastAPI apps (`mock_systems/`),
  just over an in-process ASGI transport instead of a real socket.
- For the "MCP primitives" section specifically, two servers that have *no*
  database dependency at all (`knowledge_base` and the payer mock) are booted
  as **real** `uvicorn` servers on loopback ports in background threads, so
  those calls are genuine MCP wire-protocol round trips — not faked.

## Before you run this

- Run from a Python environment that has the project's dependencies
  installed (`uv sync --group dev`, or `pip install -e .` plus the `dev`
  group) and `ipykernel`/`jupyter` available.
- No Docker, no Postgres, no network access required — everything here is
  self-contained. If your real `.env` has a working `LLMGW_API_KEY` and
  `LLM_ENABLED=true`, the LLM-backed cells (code/procedure extraction,
  specialist ranking, the chat assistant) will use the real model; if not,
  they'll deterministically fall back exactly as the app does in production
  (ADR-005) and this notebook says so explicitly at each point.
- Cells use top-level `await` (native in `ipykernel` ≥ 6 / modern
  Jupyter and VS Code notebooks).
- Run cells **in order, top to bottom, once** — later cells depend on state
  (users, referrals, an open checkpointer) created earlier.

## 1 · Setup

Locate the repo root and `chdir` into it *before* importing anything from
`app.*` — `Settings` (`app/core/config.py`) loads `.env` relative to the
current working directory, and `DATABASE_URL`/`SECRET_KEY` are required
fields with no default.

In [1]:
import os, sys
from pathlib import Path

def find_repo_root(start: Path) -> Path:
    for candidate in [start, *start.parents]:
        if (candidate / "pyproject.toml").exists() and (candidate / "app").is_dir():
            return candidate
    raise RuntimeError("Could not locate the healthcare_capstone repo root — run this notebook from within the project.")

REPO_ROOT = find_repo_root(Path.cwd())
os.chdir(REPO_ROOT)
if str(REPO_ROOT) not in sys.path:
    sys.path.insert(0, str(REPO_ROOT))

print("Repo root:", REPO_ROOT)
print(".env present:", (REPO_ROOT / ".env").exists())

Repo root: c:\Users\DELL\Documents\hcp-cp\healthcare_capstone
.env present: True


In [2]:
import asyncio
import json as _json
import logging
import io
import socket
import threading
import time
from datetime import date, timedelta, timezone, datetime as dt

from httpx import ASGITransport, AsyncClient
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession, async_sessionmaker, create_async_engine

# Importing app.main pulls in every router, model, and service — this is the
# same "one import wires everything" pattern tests/conftest.py relies on, so
# every model is registered on Base.metadata before we create tables below.
from app.main import app
from app.core.config import settings
from app.core.logging import JsonFormatter, correlation_id_var
from app.database.base import Base

print("LLM enabled (real .env):", settings.llm_enabled, "| model:", settings.llm_model if settings.llm_enabled else "(stub fallback)")

def show(obj, limit=None):
    # Pretty-print an API response / dict for the notebook output.
    text = _json.dumps(obj, indent=2, default=str)
    print(text if limit is None else text[:limit])

[08/12/26 23:13:39] INFO     No auth config provided, skipping auth setup                             ]8;id=8125617;file://c:\Users\DELL\Documents\hcp-cp\healthcare_capstone\.venv\Lib\site-packages\fastapi_mcp\server.py\server.py]8;;\:]8;id=8125618;file://c:\Users\DELL\Documents\hcp-cp\healthcare_capstone\.venv\Lib\site-packages\fastapi_mcp\server.py#310\310]8;;\

                    INFO     MCP HTTP server listening at /mcp                                        ]8;id=8125624;file://c:\Users\DELL\Documents\hcp-cp\healthcare_capstone\.venv\Lib\site-packages\fastapi_mcp\server.py\server.py]8;;\:]8;id=8125625;file://c:\Users\DELL\Documents\hcp-cp\healthcare_capstone\.venv\Lib\site-packages\fastapi_mcp\server.py#366\366]8;;\

                    INFO     No auth config provided, skipping auth setup                             ]8;id=8125630;file://c:\Users\DELL\Documents\hcp-cp\healthcare_capstone\.venv\Lib\site-packages\fastapi_mcp\server.py\server.py]8;;\:]8;id=8125631;file://c:\Users\DELL\Documents\hcp-cp\healthcare_capstone\.venv\Lib\site-packages\fastapi_mcp\server.py#310\310]8;;\

                    INFO     MCP HTTP server listening at /mcp                                        ]8;id=8125636;file://c:\Users\DELL\Documents\hcp-cp\healthcare_capstone\.venv\Lib\site-packages\fastapi_mcp\server.py\server.py]8;;\:]8;id=8125637;file://c:\Users\DELL\Documents\hcp-cp\healthcare_capstone\.venv\Lib\site-packages\fastapi_mcp\server.py#366\366]8;;\

                    INFO     No auth config provided, skipping auth setup                             ]8;id=8125642;file://c:\Users\DELL\Documents\hcp-cp\healthcare_capstone\.venv\Lib\site-packages\fastapi_mcp\server.py\server.py]8;;\:]8;id=8125643;file://c:\Users\DELL\Documents\hcp-cp\healthcare_capstone\.venv\Lib\site-packages\fastapi_mcp\server.py#310\310]8;;\

                    INFO     MCP HTTP server listening at /mcp                                        ]8;id=8125648;file://c:\Users\DELL\Documents\hcp-cp\healthcare_capstone\.venv\Lib\site-packages\fastapi_mcp\server.py\server.py]8;;\:]8;id=8125649;file://c:\Users\DELL\Documents\hcp-cp\healthcare_capstone\.venv\Lib\site-packages\fastapi_mcp\server.py#366\366]8;;\

                    INFO     No auth config provided, skipping auth setup                             ]8;id=8125654;file://c:\Users\DELL\Documents\hcp-cp\healthcare_capstone\.venv\Lib\site-packages\fastapi_mcp\server.py\server.py]8;;\:]8;id=8125655;file://c:\Users\DELL\Documents\hcp-cp\healthcare_capstone\.venv\Lib\site-packages\fastapi_mcp\server.py#310\310]8;;\

                    INFO     MCP HTTP server listening at /mcp                                        ]8;id=8125660;file://c:\Users\DELL\Documents\hcp-cp\healthcare_capstone\.venv\Lib\site-packages\fastapi_mcp\server.py\server.py]8;;\:]8;id=8125661;file://c:\Users\DELL\Documents\hcp-cp\healthcare_capstone\.venv\Lib\site-packages\fastapi_mcp\server.py#366\366]8;;\

                    INFO     No auth config provided, skipping auth setup                             ]8;id=8125666;file://c:\Users\DELL\Documents\hcp-cp\healthcare_capstone\.venv\Lib\site-packages\fastapi_mcp\server.py\server.py]8;;\:]8;id=8125667;file://c:\Users\DELL\Documents\hcp-cp\healthcare_capstone\.venv\Lib\site-packages\fastapi_mcp\server.py#310\310]8;;\

                    INFO     MCP HTTP server listening at /mcp                                        ]8;id=8125672;file://c:\Users\DELL\Documents\hcp-cp\healthcare_capstone\.venv\Lib\site-packages\fastapi_mcp\server.py\server.py]8;;\:]8;id=8125673;file://c:\Users\DELL\Documents\hcp-cp\healthcare_capstone\.venv\Lib\site-packages\fastapi_mcp\server.py#366\366]8;;\

{"ts": "2026-08-12T23:13:40+0530", "level": "INFO", "logger": "fastapi_mcp.server", "correlation_id": "-", "message": "No auth config provided, skipping auth setup"}
{"ts": "2026-08-12T23:13:40+0530", "level": "INFO", "logger": "fastapi_mcp.server", "correlation_id": "-", "message": "MCP HTTP server listening at /mcp"}
LLM enabled (real .env): True | model: openai/gpt-oss-120b


### Database

A file-backed SQLite database in a fresh temp directory — not `:memory:`
with `StaticPool` the way `tests/conftest.py` does it. That pattern keeps
exactly *one* physical connection alive for the whole engine, which is fine
for a single request at a time but deadlocks here: Section 8 below holds an
SSE stream open (its request-scoped DB session checked out for the whole
time the stream is open) *while* concurrently writing through a second
session — with only one connection to go around, the second session blocks
forever waiting for the first to give it back. A real file gives SQLite (and
this engine's connection pool) room for more than one connection at once,
exactly like pointing at real Postgres would.

In [3]:
import tempfile

_demo_db_dir = Path(tempfile.mkdtemp(prefix="capstone_demo_"))
_demo_db_path = _demo_db_dir / "demo.sqlite3"

engine = create_async_engine(
    f"sqlite+aiosqlite:///{_demo_db_path}",
    connect_args={"check_same_thread": False, "timeout": 30},
)
demo_sessionmaker = async_sessionmaker(engine, class_=AsyncSession, expire_on_commit=False)
print("Demo database file:", _demo_db_path)

async def _create_tables():
    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.create_all)

await _create_tables()
print("Tables created:", sorted(Base.metadata.tables.keys()))

Demo database file: C:\Users\DELL\AppData\Local\Temp\capstone_demo_qwggoug5\demo.sqlite3
Tables created: ['appointments', 'audit_logs', 'doctor_availability', 'doctor_insurance_networks', 'doctors', 'insurance_plans', 'medical_record_documents', 'medical_records', 'notifications', 'outbox_events', 'patients', 'permissions', 'provider_directory_links', 'referral_documents', 'referral_outcomes', 'referral_requests', 'refresh_tokens', 'role_permissions', 'roles', 'schedule_slots', 'specialist_notes', 'user_roles', 'users']


### Wiring the app onto the demo database

`get_async_session` (used by every route via `Depends(...)`) and every
LangGraph node's `db_session.async_session()` call both resolve the
`async_session` name from `app.database.session`'s own module globals *at
call time* — see the docstring on `app/services/referral_workflow.py`. One
monkeypatch here is enough for the whole app, HTTP routes and background
tasks alike; no per-request `dependency_overrides` needed.

In [4]:
import app.database.session as db_session_module
db_session_module.async_session = demo_sessionmaker
print("app.database.session.async_session -> demo_sessionmaker (in-memory SQLite)")

app.database.session.async_session -> demo_sessionmaker (in-memory SQLite)


### Seeding roles & permissions

Runs the exact same idempotent seeding function `scripts/seed_roles.py` and
the Docker entrypoint (`scripts/start.sh`) call in production.

In [5]:
from app.core.seed import seed_roles_and_permissions, ROLE_PERMISSIONS

async with demo_sessionmaker() as db:
    await seed_roles_and_permissions(db)

print("Seeded roles:", list(ROLE_PERMISSIONS.keys()))

Seeded roles: ['patient', 'pcp', 'specialist', 'care_coordinator', 'payer_admin', 'doctor', 'admin']


### Faking the MCP tool-calling seam for the referral workflow

`app.agents.mcp_clients.get_tools` is the one seam every LangGraph node calls
through to reach an MCP tool. We swap it for
`tests.agent_fakes.fake_get_tools` — the **same double the real test suite
uses** — which still executes the real `mock_systems/*` FastAPI apps, just
over an in-process ASGI transport instead of a real socket (no live server
needed for this part). The workflow logic that calls it is completely
unmodified.

In [6]:
import app.agents.mcp_clients as mcp_clients
from tests.agent_fakes import fake_get_tools

mcp_clients.get_tools = fake_get_tools
print("app.agents.mcp_clients.get_tools -> tests.agent_fakes.fake_get_tools")

app.agents.mcp_clients.get_tools -> tests.agent_fakes.fake_get_tools


### Building the referral workflow graph against a SQLite checkpointer

Production uses `AsyncPostgresSaver` (ADR-004 — workflow state externalized
so a human-in-the-loop pause survives a restart). We use
`AsyncSqliteSaver` instead — real checkpoint serialize/restore, just a
different backend, exactly what `tests/conftest.py` does. `_compiled_graph`/
`_checkpointer` are the same module globals `app.main`'s lifespan sets on a
real boot; we set them directly since lifespan never runs under a notebook.

In [7]:
from langgraph.checkpoint.sqlite.aio import AsyncSqliteSaver
import app.agents.graph as agent_graph

_checkpointer_cm = AsyncSqliteSaver.from_conn_string(":memory:")
checkpointer = await _checkpointer_cm.__aenter__()

agent_graph._compiled_graph = agent_graph.build_graph(checkpointer)
agent_graph._checkpointer = checkpointer

print("Referral workflow graph compiled against an in-memory SQLite checkpointer.")
print("Nodes:", list(agent_graph.get_compiled_graph().get_graph().nodes.keys()))

Referral workflow graph compiled against an in-memory SQLite checkpointer.
Nodes: ['__start__', 'intake', 'await_documents', 'verify_eligibility', 'escalate_eligibility', 'recommend_specialist', 'await_specialist_approval', 'schedule_appointment', 'book_real_appointment', 'summarize_for_specialist', 'notify', '__end__']


### The HTTP client

An `httpx.AsyncClient` wired directly to the FastAPI `app` object via
`ASGITransport` — every request below runs the *real* ASGI stack
(middlewares, dependency injection, routing) with no real socket, the same
mechanism `tests/conftest.py::test_client` uses.

In [8]:
client = AsyncClient(transport=ASGITransport(app=app), base_url="http://demo")

def auth(token: str) -> dict:
    return {"Authorization": f"Bearer {token}"}

print("client ready ->", client.base_url)

client ready -> http://demo


## 2 · JWT authentication — raw tokens, tampering, expiry, refresh rotation

The dashboard only ever shows you "logged in" or "logged out." Here's what's
actually inside the token, and what happens when it's altered, expired, or
replayed.

In [9]:
patient_amit_payload = {
    "email": "amit.demo@example.com",
    "username": "amit_demo",
    "password": "demoPass123!",
    "register_as_patient": True,
    "first_name": "Amit",
    "last_name": "Shah",
    "date_of_birth": "1988-04-12",
    "gender": "male",
    "phone": "+15550100",
    "insurance_provider": "Acme PPO Gold",
    "insurance_policy_number": "ACME-991123",  # a real, verified plan in mock_systems/payer_mock
}

resp = await client.post("/auth/register", json=patient_amit_payload)
print(resp.status_code)
show(resp.json())

{"ts": "2026-08-12T23:15:26+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: POST http://demo/auth/register \"HTTP/1.1 201 Created\""}
201
{
  "id": 1,
  "email": "amit.demo@example.com",
  "username": "amit_demo",
  "is_active": true,
  "is_superuser": false
}


In [10]:
patient_zara_payload = {
    "email": "zara.demo@example.com",
    "username": "zara_demo",
    "password": "demoPass123!",
    "register_as_patient": True,
    "first_name": "Zara",
    "last_name": "Ali",
    "date_of_birth": "1995-09-03",
    "gender": "female",
    "phone": "+15550101",
    "insurance_provider": "Unrecognized Plan Co",
    "insurance_policy_number": "UNKNOWN-000",  # NOT in mock_systems/payer_mock's PLANS -> denied
}

resp = await client.post("/auth/register", json=patient_zara_payload)
print(resp.status_code)
show(resp.json())

{"ts": "2026-08-12T23:15:40+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: POST http://demo/auth/register \"HTTP/1.1 201 Created\""}
201
{
  "id": 2,
  "email": "zara.demo@example.com",
  "username": "zara_demo",
  "is_active": true,
  "is_superuser": false
}


In [11]:
resp = await client.post("/auth/login", data={"username": "amit_demo", "password": "demoPass123!"})
tokens_amit = resp.json()
print(resp.status_code)
show(tokens_amit)

{"ts": "2026-08-12T23:15:45+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: POST http://demo/auth/login \"HTTP/1.1 200 OK\""}
200
{
  "access_token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJhbWl0X2RlbW8iLCJleHAiOjE3ODY1NTg1NDUsInR5cGUiOiJhY2Nlc3MifQ.MXoNaK-8bvhjafzkuAD0LaTlSIPG_geEg_HtNR8m51M",
  "refresh_token": "HZThPCwK-DWuyhOWGlVWT59nAzLHODPQ1UEIFKvQ6f6wVmjnFNn1lvUinBQ8FSjD",
  "token_type": "bearer"
}


### Decoding the token by hand

This is the part the UI never shows you: the access token is a signed JWT
carrying `sub` (username), `exp`, and `type` — nothing else. Decoding it
*without* verification (as any browser dev tools / jwt.io would) shows the
claims; decoding it *with* verification (`verify_token`, what every protected
route actually does) is what matters for security.

In [12]:
from jose import jwt as jose_jwt
from app.auth.jwt_handler import verify_token, create_access_token

raw_claims = jose_jwt.get_unverified_claims(tokens_amit["access_token"])
print("Unverified claims (what anyone can read without the secret key):")
show(raw_claims)

print("\nverify_token() [checks signature + exp against SECRET_KEY] ->", verify_token(tokens_amit["access_token"]))

Unverified claims (what anyone can read without the secret key):
{
  "sub": "amit_demo",
  "exp": 1786558545,
  "type": "access"
}

verify_token() [checks signature + exp against SECRET_KEY] -> amit_demo


In [13]:
# Tamper with one character in the signature segment.
parts = tokens_amit["access_token"].split(".")
tampered = ".".join([parts[0], parts[1], parts[2][:-1] + ("A" if parts[2][-1] != "A" else "B")])

print("verify_token() on a tampered token ->", verify_token(tampered))

resp = await client.get("/auth/me", headers=auth(tampered))
print(f"\nGET /auth/me with tampered token -> {resp.status_code}")
show(resp.json())

verify_token() on a tampered token -> None
{"ts": "2026-08-12T23:16:04+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: GET http://demo/auth/me \"HTTP/1.1 401 Unauthorized\""}

GET /auth/me with tampered token -> 401
{
  "detail": "Could not validate credentials"
}


In [14]:
# An already-expired token — expires_delta is negative.
expired_token = create_access_token(data={"sub": "amit_demo"}, expires_delta=timedelta(seconds=-1))
resp = await client.get("/auth/me", headers=auth(expired_token))
print(f"GET /auth/me with an expired token -> {resp.status_code}")
show(resp.json())

{"ts": "2026-08-12T23:16:19+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: GET http://demo/auth/me \"HTTP/1.1 401 Unauthorized\""}
GET /auth/me with an expired token -> 401
{
  "detail": "Could not validate credentials"
}


### Refresh token rotation + reuse detection

Every `/auth/refresh` call revokes the presented token and issues a brand
new one (`rotate_refresh_token`). Using the *same* refresh token twice is
exactly what a stolen/replayed token looks like — the second use is flagged
and audited, not silently accepted.

In [15]:
old_refresh = tokens_amit["refresh_token"]

resp1 = await client.post("/auth/refresh", json={"refresh_token": old_refresh})
print("First use of the refresh token ->", resp1.status_code)
new_tokens = resp1.json()
tokens_amit = new_tokens  # keep using the freshest pair for the rest of the notebook

resp2 = await client.post("/auth/refresh", json={"refresh_token": old_refresh})
print("Second use of the SAME (now-revoked) refresh token ->", resp2.status_code)
show(resp2.json())

{"ts": "2026-08-12T23:16:28+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: POST http://demo/auth/refresh \"HTTP/1.1 200 OK\""}
First use of the refresh token -> 200
{"ts": "2026-08-12T23:16:28+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: POST http://demo/auth/refresh \"HTTP/1.1 401 Unauthorized\""}
Second use of the SAME (now-revoked) refresh token -> 401
{
  "detail": "Invalid refresh token"
}


In [16]:
from app.models.audit import AuditLog

async with demo_sessionmaker() as db:
    rows = (await db.execute(
        select(AuditLog).where(AuditLog.action == "auth.refresh_token_reuse_detected")
    )).scalars().all()

print(f"{len(rows)} audit log row(s) for the reuse attempt — this is the only trace of it; nothing about it is visible in any UI.")
for r in rows:
    print(r.timestamp, r.action, r.details)

1 audit log row(s) for the reuse attempt — this is the only trace of it; nothing about it is visible in any UI.
2026-08-12 17:46:28 auth.refresh_token_reuse_detected {"resource_type": "refresh_token", "resource_id": 1}


## 3 · RBAC — permission tables, not hardcoded role strings

`require_permission(name)` (`app/api/dependencies/auth.py`) checks the
caller's *granted permissions*, loaded from the DB-driven `Role`/`Permission`
tables — never a hardcoded `if role == "admin"`. `admin:*` short-circuits
every check. Staff roles (pcp, specialist, care_coordinator, admin) aren't
self-service — they're provisioned directly, the same way
`app/api/routes/admin.py::grant_role` does it in production.

In [17]:
from app.core.security import get_password_hash
from app.models.user import User
from app.models.role import Role, UserRole
from app.models.doctor import Doctor

async def create_staff_user(username, email, password, role_name):
    async with demo_sessionmaker() as db:
        user = User(username=username, email=email, hashed_password=get_password_hash(password))
        db.add(user)
        await db.flush()
        role = (await db.execute(select(Role).where(Role.name == role_name))).scalar_one()
        db.add(UserRole(user_id=user.id, role_id=role.id))
        await db.commit()
        await db.refresh(user)
        return user

async def create_doctor(first_name, last_name, specialization, email, license_number, user_id=None, city=None):
    async with demo_sessionmaker() as db:
        doctor = Doctor(
            first_name=first_name, last_name=last_name, specialization=specialization,
            email=email, license_number=license_number, user_id=user_id, city=city,
        )
        db.add(doctor)
        await db.commit()
        await db.refresh(doctor)
        return doctor

pcp_user = await create_staff_user("dr_patel_pcp", "patel.pcp@demo.example", "demoPass123!", "pcp")
coordinator_user = await create_staff_user("coord_jane", "jane.coord@demo.example", "demoPass123!", "care_coordinator")
admin_user = await create_staff_user("admin_sam", "sam.admin@demo.example", "demoPass123!", "admin")

dr_patel = await create_doctor("Anika", "Patel", "General Medicine", "dr.patel@demo.example", "PCP-DEMO-001", user_id=pcp_user.id)
# A second real, bookable platform doctor - used later to map a mock
# directory candidate onto a real doctor via ProviderDirectoryLink.
dr_kim = await create_doctor("Daniel", "Kim", "Orthopedics", "dr.kim@demo.example", "ORTHO-DEMO-002")

print("Seeded:", pcp_user.username, coordinator_user.username, admin_user.username)
print("Doctors:", dr_patel.id, dr_patel.last_name, "|", dr_kim.id, dr_kim.last_name)

Seeded: dr_patel_pcp coord_jane admin_sam
Doctors: 1 Patel | 2 Kim


In [19]:
async def login(username, password):
    resp = await client.post("/auth/login", data={"username": username, "password": password})
    resp.raise_for_status()
    return resp.json()

tokens_pcp = await login("dr_patel_pcp", "demoPass123!")
tokens_coordinator = await login("coord_jane", "demoPass123!")
tokens_admin = await login("admin_sam", "demoPass123!")
tokens_zara = await login("zara_demo", "demoPass123!")

for label, tok in [("patient (amit)", tokens_amit), ("pcp", tokens_pcp), ("care_coordinator", tokens_coordinator), ("admin", tokens_admin)]:
    resp = await client.get("/auth/me", headers=auth(tok["access_token"]))
    me = resp.json()
    print(f"{label:20s} roles={me['roles']}  permissions={me['permissions']}")

{"ts": "2026-08-12T23:17:10+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: POST http://demo/auth/login \"HTTP/1.1 200 OK\""}
{"ts": "2026-08-12T23:17:10+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: POST http://demo/auth/login \"HTTP/1.1 200 OK\""}
{"ts": "2026-08-12T23:17:10+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: POST http://demo/auth/login \"HTTP/1.1 200 OK\""}
{"ts": "2026-08-12T23:17:11+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: POST http://demo/auth/login \"HTTP/1.1 200 OK\""}
{"ts": "2026-08-12T23:17:11+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: GET http://demo/auth/me \"HTTP/1.1 200 OK\""}
patient (amit)       roles=['patient']  permissions=['appointment:view_own', 'medical_record:manage', 'medical_record:view_own', 'patient:view_own', 'refe

### The same route, four different outcomes

`GET /referral/requests/ops-queue` requires `referral:approve`. Watch the
same endpoint behave differently purely based on the DB-driven permission
set of whoever's calling it.

In [20]:
for label, tok in [("patient", tokens_amit), ("pcp", tokens_pcp), ("care_coordinator", tokens_coordinator), ("admin", tokens_admin)]:
    resp = await client.get("/referral/requests/ops-queue", headers=auth(tok["access_token"]))
    print(f"{label:18s} -> {resp.status_code}  {'(missing referral:approve)' if resp.status_code == 403 else ''}")

{"ts": "2026-08-12T23:17:24+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: GET http://demo/referral/requests/ops-queue \"HTTP/1.1 403 Forbidden\""}


patient            -> 403  (missing referral:approve)
{"ts": "2026-08-12T23:17:24+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: GET http://demo/referral/requests/ops-queue \"HTTP/1.1 403 Forbidden\""}
pcp                -> 403  (missing referral:approve)
{"ts": "2026-08-12T23:17:24+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: GET http://demo/referral/requests/ops-queue \"HTTP/1.1 200 OK\""}
care_coordinator   -> 200  
{"ts": "2026-08-12T23:17:24+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: GET http://demo/referral/requests/ops-queue \"HTTP/1.1 200 OK\""}
admin              -> 200  


### Row-level scoping, not just route gates

Two patients both hold `referral:view_own` — same permission — but that only
ever resolves to *their own* rows (`referral_visibility_filter`, see
`app/services/referral_scope.py`). We'll come back and populate a real
referral for Amit further down; for now, confirm Zara can't enumerate
anyone else's data even with a fully valid token for her own account.

In [21]:
resp = await client.get("/referral/requests/", headers=auth(tokens_zara["access_token"]))
print("Zara's own referral list (should be empty so far) ->", resp.status_code)
show(resp.json())

{"ts": "2026-08-12T23:17:32+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: GET http://demo/referral/requests/ \"HTTP/1.1 200 OK\""}
Zara's own referral list (should be empty so far) -> 200
{
  "items": [],
  "total": 0,
  "skip": 0,
  "limit": 100,
  "next": null
}


## 4 · Security & middleware — correlation IDs, structured logs, rate limiting, audit trail

None of this is visible in the dashboard. It's infrastructure that exists
purely to make an incident traceable after the fact.

### Correlation ID propagation

`CorrelationIdMiddleware` reads (or generates) `X-Correlation-Id`, stashes it
in a `contextvars.ContextVar` for the duration of the request, and every log
line emitted while handling it — including from a LangGraph node running as
a background task — carries the same id. We attach a temporary log handler
to capture that JSON directly.

In [22]:
# A tiny diagnostic route, not part of the shipped API - added only so this
# notebook has something that logs mid-request, making the correlation id's
# propagation into ordinary logger.* calls directly observable. Every real
# route gets the exact same contextvar for free from CorrelationIdMiddleware;
# most of them just don't happen to call logging.* on their happy path since
# this app's business events go through the AuditLog table instead (see the
# audit trail below) rather than ad-hoc log lines.
@app.get("/__demo/correlation-probe", include_in_schema=False)
async def _correlation_probe():
    logging.getLogger("demo.correlation_probe").info("handling probe request")
    return {"correlation_id_seen_by_handler": correlation_id_var.get()}

log_buffer = io.StringIO()
capture_handler = logging.StreamHandler(log_buffer)
capture_handler.setFormatter(JsonFormatter())
logging.getLogger().addHandler(capture_handler)

trace_id = "demo-trace-4f9c21"
resp = await client.get("/__demo/correlation-probe", headers={"X-Correlation-Id": trace_id})
print("Response body:", resp.json())
print("Response header X-Correlation-Id:", resp.headers.get("x-correlation-id"))
print("Header, body, and the contextvar the handler saw all match what we sent:",
      resp.headers.get("x-correlation-id") == trace_id == resp.json()["correlation_id_seen_by_handler"])

logging.getLogger().removeHandler(capture_handler)
lines = [ln for ln in log_buffer.getvalue().splitlines() if ln.strip()]
matching = [_json.loads(ln) for ln in lines if _json.loads(ln).get("correlation_id") == trace_id]
print(f"\n{len(matching)} log line(s) tagged with {trace_id!r} — this is what an on-call engineer greps for:")
for m in matching:
    show(m)

{"ts": "2026-08-12T23:17:38+0530", "level": "INFO", "logger": "demo.correlation_probe", "correlation_id": "demo-trace-4f9c21", "message": "handling probe request"}
{"ts": "2026-08-12T23:17:38+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: GET http://demo/__demo/correlation-probe \"HTTP/1.1 200 OK\""}
Response body: {'correlation_id_seen_by_handler': 'demo-trace-4f9c21'}
Response header X-Correlation-Id: demo-trace-4f9c21
Header, body, and the contextvar the handler saw all match what we sent: True

1 log line(s) tagged with 'demo-trace-4f9c21' — this is what an on-call engineer greps for:
{
  "ts": "2026-08-12T23:17:38+0530",
  "level": "INFO",
  "logger": "demo.correlation_probe",
  "correlation_id": "demo-trace-4f9c21",
  "message": "handling probe request"
}


### Rate limiting

`POST /auth/login` is limited to 5/minute per client IP (`slowapi`). We reset
the in-process limiter first (it persists across calls in the same process —
see `tests/conftest.py::reset_rate_limiter` for why), then hammer it.

In [23]:
from app.core.rate_limit import limiter

limiter.reset()
statuses = []
for i in range(6):
    resp = await client.post("/auth/login", data={"username": "amit_demo", "password": "demoPass123!"})
    statuses.append(resp.status_code)
print("6 rapid login attempts ->", statuses)
print("6th request blocked by the rate limiter:", statuses[-1] == 429)
limiter.reset()  # don't let this bleed into later cells

{"ts": "2026-08-12T23:17:44+0530", "level": "INFO", "logger": "slowapi", "correlation_id": "-", "message": "Storage has been reset and all limits cleared"}
{"ts": "2026-08-12T23:17:44+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: POST http://demo/auth/login \"HTTP/1.1 200 OK\""}
{"ts": "2026-08-12T23:17:45+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: POST http://demo/auth/login \"HTTP/1.1 200 OK\""}
{"ts": "2026-08-12T23:17:45+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: POST http://demo/auth/login \"HTTP/1.1 200 OK\""}
{"ts": "2026-08-12T23:17:45+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: POST http://demo/auth/login \"HTTP/1.1 200 OK\""}
{"ts": "2026-08-12T23:17:46+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: POST http://demo/auth/login \"HTTP/1.1 200 O

### The audit trail

Every login, referral mutation, and (as we'll see in Section 6) every MCP
tool call writes an `AuditLog` row in the *same transaction* as the change it
records — never a separate commit, so there's no window where an action
happened but wasn't logged (or vice versa).

In [24]:
async with demo_sessionmaker() as db:
    rows = (await db.execute(select(AuditLog).order_by(AuditLog.id))).scalars().all()

print(f"{len(rows)} audit log rows so far:")
for r in rows[-10:]:
    print(f"  #{r.id:<4} {str(r.timestamp):26s} user_id={str(r.user_id):<5} {r.action}")

16 audit log rows so far:
  #7    2026-08-12 17:46:28        user_id=1     auth.refresh_token_reuse_detected
  #8    2026-08-12 17:47:10        user_id=3     auth.login
  #9    2026-08-12 17:47:10        user_id=4     auth.login
  #10   2026-08-12 17:47:10        user_id=5     auth.login
  #11   2026-08-12 17:47:11        user_id=2     auth.login
  #12   2026-08-12 17:47:44        user_id=1     auth.login
  #13   2026-08-12 17:47:45        user_id=1     auth.login
  #14   2026-08-12 17:47:45        user_id=1     auth.login
  #15   2026-08-12 17:47:45        user_id=1     auth.login
  #16   2026-08-12 17:47:46        user_id=1     auth.login


## 5 · MCP primitives — tools, resources, and prompts, over a real running server

`fastapi_mcp.FastApiMCP.mount_http()` (used by the platform API and every
`mock_systems/*` server) only ever converts REST routes into **tools**. The
one server in this project that speaks all three MCP primitives is
`knowledge_base/main.py`, built on the raw `mcp` SDK's `FastMCP` — it exposes
a tool, two resources, and two prompt templates over the local policy corpus
(`nb/`). We boot it for real on a loopback port (a genuine `uvicorn` process
in a background thread — no database dependency, so this is safe and fast),
then speak real MCP protocol to it with the raw client SDK, not the
LangChain wrapper the app itself uses elsewhere.

We also boot the payer mock the same way, to show a **tool**'s JSON schema
being generated live from a plain FastAPI/Pydantic route signature (Day 10 —
"tool schema design").

In [25]:
def _free_port() -> int:
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.bind(("127.0.0.1", 0))
    port = s.getsockname()[1]
    s.close()
    return port

def start_server_in_thread(asgi_app, port, lifespan="on"):
    import uvicorn
    config = uvicorn.Config(asgi_app, host="127.0.0.1", port=port, log_level="warning", lifespan=lifespan)
    server = uvicorn.Server(config)
    thread = threading.Thread(target=server.run, daemon=True)
    thread.start()
    deadline = time.time() + 10
    while time.time() < deadline:
        try:
            with socket.create_connection(("127.0.0.1", port), timeout=0.2):
                return server, thread
        except OSError:
            time.sleep(0.1)
    raise RuntimeError(f"server on port {port} did not come up in time")

_running_servers = []  # (server, thread) pairs, stopped in the cleanup cell at the end

from knowledge_base.main import mcp as kb_mcp_server
kb_port = _free_port()
kb_server, kb_thread = start_server_in_thread(kb_mcp_server.streamable_http_app(), kb_port)
_running_servers.append((kb_server, kb_thread))
print(f"knowledge_base MCP server live on http://127.0.0.1:{kb_port}/mcp")

from mock_systems.payer_mock.main import app as payer_mock_app
payer_port = _free_port()
payer_server, payer_thread = start_server_in_thread(payer_mock_app, payer_port)
_running_servers.append((payer_server, payer_thread))
print(f"payer mock MCP server live on http://127.0.0.1:{payer_port}/mcp")

{"ts": "2026-08-12T23:18:05+0530", "level": "INFO", "logger": "mcp.server.streamable_http_manager", "correlation_id": "-", "message": "StreamableHTTP session manager started"}
knowledge_base MCP server live on http://127.0.0.1:56218/mcp
payer mock MCP server live on http://127.0.0.1:56226/mcp


In [26]:
from mcp import ClientSession
from mcp.client.streamable_http import streamablehttp_client

async def mcp_call_tool(base_url, tool_name, arguments):
    async with streamablehttp_client(f"{base_url}/mcp") as (read, write, _):
        async with ClientSession(read, write) as session:
            await session.initialize()
            result = await session.call_tool(tool_name, arguments)
            return [getattr(block, "text", block) for block in result.content]

async def mcp_list_tools(base_url):
    async with streamablehttp_client(f"{base_url}/mcp") as (read, write, _):
        async with ClientSession(read, write) as session:
            await session.initialize()
            result = await session.list_tools()
            return result.tools

### Tool discovery — real, live schema generation

The `check_eligibility` tool's JSON schema below did not come from a hand
written spec — `fastapi_mcp` generated it from the Pydantic
`EligibilityRequest` model on `mock_systems/payer_mock/main.py`'s route
signature, live, the moment the server started.

In [27]:
payer_base = f"http://127.0.0.1:{payer_port}"
tools = await mcp_list_tools(payer_base)
for t in tools:
    print(f"tool: {t.name}\n  description: {t.description}\n  input schema:")
    show(t.inputSchema)

{"ts": "2026-08-12T23:18:20+0530", "level": "INFO", "logger": "mcp.server.streamable_http_manager", "correlation_id": "-", "message": "StreamableHTTP session manager started"}
{"ts": "2026-08-12T23:18:20+0530", "level": "INFO", "logger": "fastapi_mcp.transport.http", "correlation_id": "-", "message": "StreamableHTTP session manager is running"}
{"ts": "2026-08-12T23:18:20+0530", "level": "INFO", "logger": "mcp.server.streamable_http_manager", "correlation_id": "-", "message": "Created new transport with session ID: 474f14c14f8e4eab8db07a3597dd5768"}
{"ts": "2026-08-12T23:18:20+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: POST http://127.0.0.1:56226/mcp \"HTTP/1.1 200 OK\""}
{"ts": "2026-08-12T23:18:20+0530", "level": "INFO", "logger": "mcp.client.streamable_http", "correlation_id": "-", "message": "Received session ID: 474f14c14f8e4eab8db07a3597dd5768"}
{"ts": "2026-08-12T23:18:20+0530", "level": "INFO", "logger": "mcp.client.streamable_ht

In [28]:
verified = await mcp_call_tool(payer_base, "check_eligibility", {"insurance_policy_number": "ACME-991123", "procedure_code": "M5126"})
print("check_eligibility(ACME-991123, M51.26) ->")
show(_json.loads(verified[0]))

denied = await mcp_call_tool(payer_base, "check_eligibility", {"insurance_policy_number": "UNKNOWN-000", "procedure_code": "M5126"})
print("\ncheck_eligibility(UNKNOWN-000, M51.26) ->")
show(_json.loads(denied[0]))

{"ts": "2026-08-12T23:18:45+0530", "level": "INFO", "logger": "mcp.server.streamable_http_manager", "correlation_id": "-", "message": "Created new transport with session ID: 2fff0d3891da42cc80eeefc426ffd315"}
{"ts": "2026-08-12T23:18:45+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: POST http://127.0.0.1:56226/mcp \"HTTP/1.1 200 OK\""}
{"ts": "2026-08-12T23:18:45+0530", "level": "INFO", "logger": "mcp.client.streamable_http", "correlation_id": "-", "message": "Received session ID: 2fff0d3891da42cc80eeefc426ffd315"}
{"ts": "2026-08-12T23:18:45+0530", "level": "INFO", "logger": "mcp.client.streamable_http", "correlation_id": "-", "message": "Negotiated protocol version: 2025-11-25"}
{"ts": "2026-08-12T23:18:45+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: POST http://127.0.0.1:56226/mcp \"HTTP/1.1 202 Accepted\""}
{"ts": "2026-08-12T23:18:45+0530", "level": "INFO", "logger": "mcp.server.lowlevel.se

### Resources — addressable, readable data, not a callable action

`kb://policies` is a catalog resource; `kb://policies/{doc_id}` is a
templated resource for one document's full text. Neither is a "tool call" —
these are read like a URI, the MCP resource primitive specifically.

In [29]:
from pydantic import AnyUrl

kb_base = f"http://127.0.0.1:{kb_port}"

async with streamablehttp_client(f"{kb_base}/mcp") as (read, write, _):
    async with ClientSession(read, write) as session:
        await session.initialize()

        resources = await session.list_resources()
        print("Static resources:", [r.uri for r in resources.resources])

        templates = await session.list_resource_templates()
        print("Resource templates:", [t.uriTemplate for t in templates.resourceTemplates])

        catalog = await session.read_resource(AnyUrl("kb://policies"))
        catalog_json = _json.loads(catalog.contents[0].text)
        print(f"\nkb://policies catalog ({len(catalog_json)} documents):")
        show(catalog_json)

        first_doc_id = catalog_json[0]["doc_id"]
        doc = await session.read_resource(AnyUrl(f"kb://policies/{first_doc_id}"))
        print(f"\nFull text of kb://policies/{first_doc_id} (first 300 chars):")
        print(doc.contents[0].text[:300])

{"ts": "2026-08-12T23:18:52+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: POST http://127.0.0.1:56218/mcp \"HTTP/1.1 200 OK\""}
{"ts": "2026-08-12T23:18:52+0530", "level": "INFO", "logger": "mcp.server.streamable_http", "correlation_id": "-", "message": "Terminating session: None"}
{"ts": "2026-08-12T23:18:52+0530", "level": "INFO", "logger": "mcp.client.streamable_http", "correlation_id": "-", "message": "Negotiated protocol version: 2025-11-25"}
{"ts": "2026-08-12T23:18:52+0530", "level": "ERROR", "logger": "asyncio", "correlation_id": "-", "message": "Exception in callback _ProactorBasePipeTransport._call_connection_lost(None)\nhandle: <Handle _ProactorBasePipeTransport._call_connection_lost(None)>", "exc_info": "Traceback (most recent call last):\n  File \"C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.11_3.11.2544.0_x64__qbz5n2kfra8p0\\Lib\\asyncio\\events.py\", line 84, in _run\n    self._context.run(self._callback,

### Prompts — reusable, parameterized prompt templates

`explain_referral_process` and `compare_policies` are MCP **prompt**
primitives — the platform's assistant fetches these live
(`_fetch_kb_prompt_guidance` in `app/agents/assistant_graph.py`) and folds
them into its own system prompt, rather than hardcoding the same instructions
twice.

In [30]:
async with streamablehttp_client(f"{kb_base}/mcp") as (read, write, _):
    async with ClientSession(read, write) as session:
        await session.initialize()

        prompts = await session.list_prompts()
        for p in prompts.prompts:
            print(f"prompt: {p.name}  args: {[a.name for a in (p.arguments or [])]}")

        explain = await session.get_prompt("explain_referral_process")
        print("\nexplain_referral_process ->")
        print(explain.messages[0].content.text)

        compare = await session.get_prompt("compare_policies", {"policy_a": "Acme PPO Gold", "policy_b": "Horizon Blue PPO"})
        print("\ncompare_policies(Acme PPO Gold, Horizon Blue PPO) ->")
        print(compare.messages[0].content.text)

{"ts": "2026-08-12T23:19:00+0530", "level": "INFO", "logger": "mcp.server.streamable_http", "correlation_id": "-", "message": "Terminating session: None"}
{"ts": "2026-08-12T23:19:00+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP Request: POST http://127.0.0.1:56218/mcp \"HTTP/1.1 200 OK\""}
{"ts": "2026-08-12T23:19:00+0530", "level": "INFO", "logger": "mcp.client.streamable_http", "correlation_id": "-", "message": "Negotiated protocol version: 2025-11-25"}
{"ts": "2026-08-12T23:19:00+0530", "level": "ERROR", "logger": "asyncio", "correlation_id": "-", "message": "Exception in callback _ProactorBasePipeTransport._call_connection_lost(None)\nhandle: <Handle _ProactorBasePipeTransport._call_connection_lost(None)>", "exc_info": "Traceback (most recent call last):\n  File \"C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.11_3.11.2544.0_x64__qbz5n2kfra8p0\\Lib\\asyncio\\events.py\", line 84, in _run\n    self._context.run(self._callback,

## 6 · The LangGraph referral workflow — service/workflow triggering, real state, real pauses

`POST /referral/requests/` returns `202 Accepted` immediately and hands the
actual work to a `BackgroundTask` (`run_referral_workflow`) that drives the
*same* compiled graph production uses: `intake -> verify_eligibility ->
recommend_specialist -> await_specialist_approval (interrupt) ->
schedule_appointment -> summarize_for_specialist -> notify`. Nothing about
this is visible from the dashboard beyond a status string changing.

In [31]:
async with demo_sessionmaker() as db:
    from app.models.patient import Patient
    amit_patient = (await db.execute(select(Patient).where(Patient.email == "amit.demo@example.com"))).scalar_one()
    zara_patient = (await db.execute(select(Patient).where(Patient.email == "zara.demo@example.com"))).scalar_one()

resp = await client.post(
    "/referral/requests/",
    json={
        "patient_id": amit_patient.id,
        "referring_doctor_id": dr_patel.id,
        "request_date": str(date.today()),
        "reason": "Persistent right knee pain after a fall three weeks ago, suspect meniscus tear.",
        "target_wait_days": 10,
    },
    headers=auth(tokens_pcp["access_token"]),
)
print(resp.status_code)
referral_1 = resp.json()
show(referral_1)
rid_1 = referral_1["id"]

{"ts": "2026-08-12T23:19:37+0530", "level": "INFO", "logger": "httpx", "correlation_id": "19c91ac4089a4e899574620fd8262f3d", "message": "HTTP Request: POST https://api.groq.com/openai/v1/chat/completions \"HTTP/1.1 200 OK\""}
{"ts": "2026-08-12T23:19:37+0530", "level": "INFO", "logger": "httpx", "correlation_id": "19c91ac4089a4e899574620fd8262f3d", "message": "HTTP Request: POST http://mock/eligibility/check \"HTTP/1.1 200 OK\""}
{"ts": "2026-08-12T23:19:38+0530", "level": "INFO", "logger": "httpx", "correlation_id": "19c91ac4089a4e899574620fd8262f3d", "message": "HTTP Request: GET http://mock/providers/search?specialty=Orthopedics \"HTTP/1.1 200 OK\""}
{"ts": "2026-08-12T23:19:41+0530", "level": "INFO", "logger": "httpx", "correlation_id": "19c91ac4089a4e899574620fd8262f3d", "message": "HTTP Request: POST https://api.groq.com/openai/v1/chat/completions \"HTTP/1.1 200 OK\""}
{"ts": "2026-08-12T23:19:41+0530", "level": "INFO", "logger": "httpx", "correlation_id": "-", "message": "HTTP R

### Inspecting workflow state directly — something no route fully exposes

`GET /referral-workflow/{id}/state` surfaces the LangGraph fields that aren't
columns on `ReferralRequest` (candidates, extracted codes, eligibility
result). But *whether the graph is currently paused on a human decision* —
`snapshot.interrupts` — isn't exposed by any GET route at all; the
resume/override endpoints check it internally. We read the checkpoint
directly, the same way those routes do.

In [ ]:
resp = await client.get(f"/referral-workflow/{rid_1}/state", headers=auth(tokens_pcp["access_token"]))
print("GET /referral-workflow/{id}/state ->")
show(resp.json())

In [32]:
graph = agent_graph.get_compiled_graph()
config_1 = {"configurable": {"thread_id": f"referral-{rid_1}"}}
snapshot = await graph.aget_state(config_1)

print("Next node(s) to run:", snapshot.next)
print("\nPending interrupt payload (this is what a coordinator's UI would need to render, but there is no GET route for it):")
show([i.value for i in snapshot.interrupts])

Next node(s) to run: ('await_specialist_approval',)

Pending interrupt payload (this is what a coordinator's UI would need to render, but there is no GET route for it):
[
  {
    "candidates": [
      {
        "doctor_id": 88,
        "name": "Dr. Priya Rao",
        "specialty": "Orthopedics",
        "location": "Downtown",
        "distance_mi": 0.9,
        "rating": 4.8,
        "next_available_days": 3,
        "in_network": false,
        "score": 0.83,
        "reasons": [
          "Short distance (0.9\u202fmi)",
          "High rating (4.8)",
          "Available in 3\u202fdays"
        ]
      },
      {
        "doctor_id": 91,
        "name": "Dr. Daniel Kim",
        "specialty": "Orthopedics",
        "location": "Downtown",
        "distance_mi": 2.1,
        "rating": 4.6,
        "next_available_days": 5,
        "in_network": false,
        "score": 0.45,
        "reasons": [
          "Moderate distance (2.1\u202fmi)",
          "Good rating (4.6)",
          "Wait

### Auditing the eligibility MCP call — with PII redacted

`call_tool_audited` (`app/agents/audit.py`) redacts `insurance_policy_number`
before it ever reaches a log line. Confirm that directly against the audit
row this exact referral just produced.

In [33]:
async with demo_sessionmaker() as db:
    rows = (await db.execute(
        select(AuditLog).where(AuditLog.action == "agent.tool_call:check_eligibility").order_by(AuditLog.id.desc()).limit(1)
    )).scalars().all()

print("Most recent check_eligibility audit row:")
for r in rows:
    show(_json.loads(r.details))
print("\ninsurance_policy_number was ACME-991123 on the wire — note it never appears in the audit log above.")

Most recent check_eligibility audit row:
{
  "resource_type": "referral_request",
  "resource_id": 1,
  "args": {
    "insurance_policy_number": "***",
    "procedure_code": "M54.5"
  },
  "actor": "agent"
}

insurance_policy_number was ACME-991123 on the wire — note it never appears in the audit log above.


### Resuming the human-in-the-loop pause

`recommend_specialist` already returned ranked candidates from the mock
provider directory (synthetic ids like 88/91/...). We approve one **and**
map it onto a real, bookable platform doctor (`dr_kim`) via the optional
`platform_doctor_id` field — this creates a `ProviderDirectoryLink` row and
is what actually lets `book_slot`'s appointment attach to a real `doctors`
row.

In [34]:
candidates = snapshot.interrupts[0].value["candidates"]
print("Candidates presented to the coordinator:")
show(candidates)
chosen = candidates[0]

Candidates presented to the coordinator:
[
  {
    "doctor_id": 88,
    "name": "Dr. Priya Rao",
    "specialty": "Orthopedics",
    "location": "Downtown",
    "distance_mi": 0.9,
    "rating": 4.8,
    "next_available_days": 3,
    "in_network": false,
    "score": 0.83,
    "reasons": [
      "Short distance (0.9\u202fmi)",
      "High rating (4.8)",
      "Available in 3\u202fdays"
    ]
  },
  {
    "doctor_id": 91,
    "name": "Dr. Daniel Kim",
    "specialty": "Orthopedics",
    "location": "Downtown",
    "distance_mi": 2.1,
    "rating": 4.6,
    "next_available_days": 5,
    "in_network": false,
    "score": 0.45,
    "reasons": [
      "Moderate distance (2.1\u202fmi)",
      "Good rating (4.6)",
      "Wait of 5\u202fdays"
    ]
  },
  {
    "doctor_id": 73,
    "name": "Dr. Omar Farouk",
    "specialty": "Orthopedics",
    "location": "Uptown",
    "distance_mi": 4.2,
    "rating": 4.1,
    "next_available_days": 1,
    "in_network": false,
    "score": 0.33,
    "reasons"

In [ ]:
resp = await client.post(
    f"/referral-workflow/{rid_1}/resume",
    json={"doctor_id": chosen["doctor_id"], "platform_doctor_id": dr_kim.id},
    headers=auth(tokens_coordinator["access_token"]),
)
print(resp.status_code)
show(resp.json())

In [ ]:
resp = await client.get(f"/referral/requests/{rid_1}", headers=auth(tokens_pcp["access_token"]))
final_referral_1 = resp.json()
print("Final referral status:", final_referral_1["status"])

resp = await client.get(f"/referral-workflow/provider-links", headers=auth(tokens_coordinator["access_token"]))
print("\nProviderDirectoryLink rows created so far:")
show(resp.json())

resp = await client.get(f"/referral/requests/{rid_1}/notes", headers=auth(tokens_pcp["access_token"]))
print("\nAI-generated (or template-fallback) specialist note:")
show(resp.json())

### A second, independent human-in-the-loop pause — the eligibility-denial path

Zara's insurance policy (`UNKNOWN-000`) isn't in the mock payer's `PLANS`
table, so `eligibility_node` denies coverage and `escalate_eligibility_node`
pauses the graph *again* — a second, separate `interrupt()` with its own
resume route (`/override-eligibility`), gated on `referral:override` rather
than `referral:approve` (bypassing a failed eligibility check is a bigger
call than approving a candidate the workflow itself surfaced).

In [ ]:
resp = await client.post(
    "/referral/requests/",
    json={
        "patient_id": zara_patient.id,
        "referring_doctor_id": dr_patel.id,
        "request_date": str(date.today()),
        "reason": "New-onset chest pain with palpitations on exertion, rule out cardiac cause.",
        "target_wait_days": 7,
    },
    headers=auth(tokens_pcp["access_token"]),
)
referral_2 = resp.json()
rid_2 = referral_2["id"]
print(resp.status_code, "-> referral", rid_2)

config_2 = {"configurable": {"thread_id": f"referral-{rid_2}"}}
snapshot_2 = await graph.aget_state(config_2)
print("Status after intake+eligibility:", (await client.get(f'/referral/requests/{rid_2}', headers=auth(tokens_pcp['access_token']))).json()["status"])
print("Paused, next node:", snapshot_2.next)
show(snapshot_2.interrupts[0].value)

In [ ]:
resp = await client.post(
    f"/referral-workflow/{rid_2}/override-eligibility",
    json={"comment": "Manually verified coverage with the payer by phone — proceeding."},
    headers=auth(tokens_coordinator["access_token"]),
)
print("override-eligibility ->", resp.status_code, resp.json())

snapshot_2 = await graph.aget_state(config_2)
print("Paused again, next node:", snapshot_2.next, "(this time: specialist approval, same as referral 1)")
cardio_candidates = snapshot_2.interrupts[0].value["candidates"]
show(cardio_candidates)

In [ ]:
resp = await client.post(
    f"/referral-workflow/{rid_2}/resume",
    json={"doctor_id": cardio_candidates[0]["doctor_id"]},  # no platform_doctor_id this time - stays a mock-directory reference
    headers=auth(tokens_coordinator["access_token"]),
)
print(resp.status_code, resp.json())

resp = await client.get(f"/referral/requests/{rid_2}/timeline", headers=auth(tokens_coordinator["access_token"]))
print("\nFull milestone timeline for referral 2 (built entirely from the durable outbox_events trail):")
show(resp.json())

### Checkpoint history — every intermediate state, not just the latest

`aget_state_history` walks every checkpoint LangGraph ever wrote for this
thread — the full sequence of state transitions, something genuinely
impossible to reconstruct from the REST API alone.

In [ ]:
history = [s async for s in graph.aget_state_history(config_1)]
print(f"{len(history)} checkpoints recorded for referral {rid_1}:")
for snap in reversed(history):
    print(f"  next={str(snap.next):32s} status={snap.values.get('status')}")

## 7 · Graph of architecture

Two views: the LangGraph workflow's *actual* compiled topology (introspected
straight from the running graph object, not hand-drawn), and the system
architecture around it.

In [ ]:
mermaid_source = agent_graph.get_compiled_graph().get_graph().draw_mermaid()
print(mermaid_source)

In [ ]:
try:
    from IPython.display import Markdown, display
    display(Markdown(f"```mermaid\n{mermaid_source}\n```"))
except Exception as exc:
    print("(Inline mermaid rendering not available in this frontend - the raw source above renders fine on GitHub, mermaid.live, or VS Code's markdown preview.)")
    print(exc)

### System architecture

```mermaid
flowchart TB
    UI["Dashboard (static/, served at /app)\nor this notebook"] -->|"HTTPS + JWT bearer"| MW

    subgraph FastAPI["FastAPI app (app/main.py)"]
        MW["Middlewares\nCORS -> CorrelationId -> NoCacheDashboard -> RateLimiter"]
        MW --> Routers["Routers\nauth, patients, doctors, appointments, medical_records,\nreferral, schedule, analytics, admin, notifications,\nai/referral_workflow, ai/assistant"]
        Routers --> RBAC["get_current_active_user / require_permission()\napp/api/dependencies/auth.py"]
        RBAC --> Services["Services\nrecord_scope, referral_scope, audit,\nnotifications, eligibility, doctor_recommendation"]
    end

    Services --> DB[("Postgres\napp tables + LangGraph checkpoints")]

    Routers --> Graph["LangGraph referral workflow\napp/agents/graph.py"]
    Routers --> Assistant["LangGraph ReAct assistant\napp/agents/assistant_graph.py"]

    Graph --> MCPClients["Scoped MCP clients, least-privilege per node\napp/agents/mcp_clients.py"]
    Assistant --> MCPClients2["Platform + knowledge-base MCP clients"]

    MCPClients -->|"streamable HTTP"| Mocks
    MCPClients2 -->|"streamable HTTP"| PlatformMCP["/mcp\nthis app's own API, exposed as MCP tools"]
    MCPClients2 -->|"streamable HTTP"| KBMCP["/kb/mcp\nknowledge_base FastMCP - tools + resources + prompts"]

    subgraph Mocks["mock_systems/* - each its own MCP server"]
        Payer["/mock/payer\ncheck_eligibility"]
        Directory["/mock/directory\nsearch_providers"]
        Scheduling["/mock/scheduling\nget_availability, book_slot"]
        Notify["/mock/notification\nsend_notification"]
        EHR["/mock/ehr"]
    end

    Services -->|"write_outbox_event\n(same transaction)"| Outbox[("outbox_events table")]
    Outbox --> Publisher["publish_loop\napp/events/publisher.py - polls every 0.5s"]
    Publisher --> Broadcaster["in-process broadcaster\napp/events/broadcaster.py"]
    Broadcaster -->|"Server-Sent Events"| UI
```

## 8 · Real-time status updates — the outbox → publisher → broadcaster → SSE chain

The dashboard just shows a status badge changing. Underneath: a durable
`outbox_events` row (written in the *same transaction* as the state change,
per ADR — no dual-write problem), a background poller
(`publish_loop`/`process_pending_events`, normally running every 0.5s, which
we drive manually here since the real lifespan loop doesn't run under a
notebook), an in-process pub/sub fan-out, and a live SSE stream a browser
tab would have open.

We open a stream for referral 1, record an outcome (which writes a
`referral.completed` outbox event), flush the outbox by hand, and show the
literal bytes that would have crossed the wire to a listening browser tab.

`httpx.ASGITransport` (used everywhere else in this notebook) turns out not
to work for this one call: reading its source shows it drives the whole ASGI
app to completion and buffers the *entire* response body before it ever
returns control to the caller — fine for ordinary requests, but this SSE
endpoint's generator never completes on its own (it awaits the next event
forever), so a stream opened over `ASGITransport` never returns at all. A
real socket does support incremental delivery, so this one demo boots the
same `app` object as a real `uvicorn` server (lifespan off — we already did
that setup by hand above) on a loopback port, exactly like the MCP servers
in Section 5.

In [ ]:
from app.events.publisher import process_pending_events

demo_server_port = _free_port()
demo_server, demo_server_thread = start_server_in_thread(app, demo_server_port, lifespan="off")
_running_servers.append((demo_server, demo_server_thread))
sse_client = AsyncClient(base_url=f"http://127.0.0.1:{demo_server_port}")

captured_events = []

async def _record_sse_events(response):
    async for line in response.aiter_lines():
        if line.startswith("data:"):
            captured_events.append(line[len("data:"):].strip())

async def _drive_sse_demo():
    async with sse_client.stream(
        "GET", f"/referral/requests/{rid_1}/events", headers=auth(tokens_coordinator["access_token"])
    ) as response:
        reader_task = asyncio.create_task(_record_sse_events(response))
        await asyncio.sleep(0.3)  # let the stream actually open before we cause an event

        await process_pending_events()  # flush anything already pending

        resp = await client.post(
            f"/referral/requests/{rid_1}/outcome",
            json={
                "symptoms": "Improved range of motion, mild residual swelling.",
                "diagnosis": "Partial medial meniscus tear, managed conservatively.",
                "prescription": "Continue physiotherapy 2x/week for 4 weeks.",
                "follow_up_notes": "Re-evaluate in 6 weeks; MRI if no improvement.",
            },
            headers=auth(tokens_coordinator["access_token"]),
        )
        print("POST .../outcome ->", resp.status_code)

        published = await process_pending_events()
        print(f"process_pending_events() flushed {published} outbox row(s)")

        await asyncio.sleep(0.5)  # give the broadcaster a moment to deliver it
        reader_task.cancel()
        try:
            await reader_task
        except asyncio.CancelledError:
            pass

try:
    await asyncio.wait_for(_drive_sse_demo(), timeout=15)
except asyncio.TimeoutError:
    print("SSE demo didn't finish within 15s - skipping (this is exploratory infrastructure, not core to the platform).")
finally:
    await sse_client.aclose()

print(f"\n{len(captured_events)} SSE message(s) captured on the open stream:")
for ev in captured_events:
    show(_json.loads(ev))

## 9 · The conversational assistant — role-scoped tools, real ReAct reasoning

`build_assistant_graph` (`app/agents/assistant_graph.py`) picks a
per-role tool allowlist and system prompt, then wires a real
`langgraph.prebuilt.create_react_agent` to the platform's own MCP tool
surface — authenticated as the *specific calling user*, not a shared
identity, so a tool call the model makes is exactly as scoped as if the user
had called the REST API directly. Its real MCP hop was already demonstrated
live in Section 5; here we reuse the exact same role-scoping and prompts,
wired to our in-process API client (same event loop, no extra server needed)
to actually run a conversation.

In [ ]:
from app.agents.assistant_graph import ROLE_TOOL_ALLOWLIST, ROLE_SYSTEM_PROMPTS, resolve_role_for_tools, BASE_REFERRAL_TOOLS

print("Tool allowlist per role (this is enforced in code, before an LLM ever sees a tool list):")
for role, tools_allowed in ROLE_TOOL_ALLOWLIST.items():
    print(f"  {role:18s} {sorted(tools_allowed)}")

print("\npayer_admin deliberately excludes patient/appointment/medical-record tools even though nothing")
print("stops the route from existing - it's scoped to match the role's real *permission* set:")
print(" ", sorted(ROLE_TOOL_ALLOWLIST['payer_admin']))

In [ ]:
role = resolve_role_for_tools({"patient"})
print(f"resolve_role_for_tools({{'patient'}}) -> {role!r}")
print("\nSystem prompt used for this role:\n")
print(ROLE_SYSTEM_PROMPTS[role])

In [ ]:
from app.agents.llm import StubChatModel, get_chat_model
from app.api.routes.ai.assistant import faq_fallback

llm = get_chat_model("assistant")

if isinstance(llm, StubChatModel):
    print("No LLM configured (LLM_ENABLED/LLMGW_API_KEY unset in .env) - this is the exact branch")
    print("app/api/routes/ai/assistant.py::chat() takes in production: a deterministic FAQ fallback.\n")
    for q in ["What's my referral status?", "How do I upload a document?", "When will my appointment be scheduled?"]:
        print(f"Q: {q}\nA: {faq_fallback(q)}\n")
else:
    print(f"Real LLM configured ({settings.llm_model}) - running an actual ReAct tool-calling turn.\n")

    from langchain_core.tools import StructuredTool
    from pydantic import BaseModel as PydanticBaseModel
    from langgraph.prebuilt import create_react_agent

    class _GetReferralArgs(PydanticBaseModel):
        referral_id: int

    async def _get_referral(referral_id: int) -> dict:
        resp = await client.get(f"/referral/requests/{referral_id}", headers=auth(tokens_amit["access_token"]))
        resp.raise_for_status()
        return resp.json()

    class _ListReferralsArgs(PydanticBaseModel):
        pass

    async def _list_referrals() -> dict:
        resp = await client.get("/referral/requests/", headers=auth(tokens_amit["access_token"]))
        resp.raise_for_status()
        return resp.json()

    demo_tools = [
        StructuredTool.from_function(coroutine=_get_referral, name="get_referral", args_schema=_GetReferralArgs,
                                      description="Get a single referral by id, scoped to what the caller may see."),
        StructuredTool.from_function(coroutine=_list_referrals, name="list_referrals", args_schema=_ListReferralsArgs,
                                      description="List referrals visible to the caller."),
    ]
    assert {t.name for t in demo_tools} <= BASE_REFERRAL_TOOLS

    patient_agent = create_react_agent(llm, demo_tools, checkpointer=checkpointer, prompt=ROLE_SYSTEM_PROMPTS["patient"])
    chat_config = {"configurable": {"thread_id": "demo-chat-amit-1"}}
    result = await patient_agent.ainvoke({"messages": [("user", f"What's the status of my referral #{rid_1}?")]}, config=chat_config)
    print("Assistant reply:\n")
    print(result["messages"][-1].content)

## 10 · Wrap-up

What this notebook exercised, all against real project code:

- **JWT** — token issuance, unverified vs. verified decoding, tampering,
  expiry, refresh rotation, and reuse detection.
- **RBAC** — permission-table-driven route gating and row-level visibility
  scoping (`referral_visibility_filter`).
- **Security & middleware** — correlation-id-tagged structured JSON logs,
  rate limiting, and the transactional audit trail (with PII redaction).
- **MCP primitives** — tools, resources, *and* prompts, over a real running
  MCP server (`knowledge_base`), plus live tool-schema generation
  (`payer_mock`).
- **LangGraph / service-workflow triggering** — the full referral workflow,
  two independent human-in-the-loop pauses (specialist approval,
  eligibility escalation), direct checkpoint/state introspection, and
  checkpoint history.
- **Graph of architecture** — the compiled workflow graph's real topology,
  plus the system-level architecture around it.
- **Real-time status** — the outbox → publisher → broadcaster → SSE chain,
  captured as raw wire bytes.
- **Chat bot** — role-scoped tool allowlisting and, when a real LLM is
  configured, an actual ReAct tool-calling conversation.

See `CURRICULUM_VS_SHIPPED.md` at the repo root for what's *not* covered
here because it isn't built yet (a production-wired RAG pipeline, CI/CD,
OAuth for MCP, cloud/Kubernetes deployment).

In [ ]:
for server, thread in _running_servers:
    server.should_exit = True
for server, thread in _running_servers:
    thread.join(timeout=5)

await _checkpointer_cm.__aexit__(None, None, None)
await client.aclose()
await engine.dispose()

import shutil
shutil.rmtree(_demo_db_dir, ignore_errors=True)

print("Background MCP servers stopped, checkpointer closed, HTTP client and DB engine disposed, temp DB removed.")